src/tests/tests/utils.js

```diff
@@ -467,4 +467,56 @@ QUnit.test("getHostFromDomainInput", assert => {
   );
 });
 
+// Tests algorithm used in the pixel tracking heuristic
+// It should return a common substring between two given values
+QUnit.test("findCommonSubstrings", assert => {
+
+  assert.equal(
+    utils.findCommonSubstrings('www.google.com', 'www.google.com/analytics/')[0].length >= 8,
+    true,
+    "returned matches have length greater than or equal to 8"
+  );
+
+  assert.equal(
+    utils.findCommonSubstrings('foobar.com/foo/fizz/buzz/bar', 'foobar.com/foo/bizz/fuzz/bar')[0],
+    'foobar.com/foo/',
+    "returns longest matching value from the pair of URLs"
+  );
+
+  assert.equal(
+    utils.findCommonSubstrings('www.google.com', '.w..wgoogle.analytics.com')[0] === undefined,
+    true,
+    "returns empty array if there is no common substring even in very similar string values"
+  );
+});
+
+// used in pixel tracking heuristic, given a string the estimateMaxEntropy function
+// will return the estimated entropy value from it, based on logic parsing the string's length,
+// and classes of character complication included in the string
+QUnit.test("estimateMaxEntropy", assert => {
+  assert.equal(
+    utils.estimateMaxEntropy("google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/anal") === 257,
+    true,
+    "returns length of string if it's above 256 (MAX_LS_LEN_FOR_ENTROPY_EST)"
+  );
+
+  assert.equal(
+    utils.estimateMaxEntropy("google.com/analytics") === utils.estimateMaxEntropy("GOOGLE.COM/ANALYTICS"),
+    true,
+    "if the same string is all lower case or all upper case, the returned extimated entropy value is the same"
+  );
+
+  assert.notEqual(
+    utils.estimateMaxEntropy('analytics.GOOGLE1234_') === utils.estimateMaxEntropy('ANALYTICS.google1234'),
+    true,
+    "two nearly identical strings of mixed character classes and different cases will return different values"
+  );
+
+  assert.notEqual(
+    utils.estimateMaxEntropy('google.com/analytics') === utils.estimateMaxEntropy('0191/_-google9fjkelo'),
+    true,
+    "strings of the same length but from different character classes will estimate different entropy values"
+  );
+});
```

Comment:
Missing assertion(s) for the "If all the characters come from one of these common character groups, assume that the group is the domain of possible characters." block. (Right now you can delete that block entirely, and the tests will still pass.)

Reply:
while writing tests for this block I discovered that characters in the `SEP` separator value set aren't ignored like they're supposed to be, causing strings like `google.analytics/123` and `google.ana.lytics/123` to estimate to different entropy values. I confirmed with Bennett that those two test strings should estimate to the same value. I suspect that optimizing the entropy estimating algo is beyond the scope of this PR, so I opened a new issue #2436 to tackle after this PR